### PR TITLE
Port CellConservativeQuartic to C++ with GPU support

### PR DESCRIPTION
**Pull request by wyphan**
## Summary
CellConservativeQuartic interpolater has been ported to C++ with GPU support.
Errors for CPU-only version on local with AmrLevel SingleVortex 2D and 3D are exactly 0.
Errors for GPU-enabled version on Gigan with AmrLevel SingleVortex 2D and 3D are on the order of 1e-14.

## Additional background
I have also simplified the logic for the temporary variables `ftmp`, `ctmp`, `ctmp2`.

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

**Comment by WeiqunZhang** [comment]
We need to move the following lines out of `#ifndef BL_NO_FORT`.
```
extern AMREX_EXPORT CellConservativeProtected protected_interp;
extern AMREX_EXPORT CellConservativeQuartic   quartic_interp;
```

**Comment by wyphan** [comment]
Thanks for catching that!

File: Src/AmrCore/AMReX_Interpolater.cpp [hunk 1]
```diff
@@ -729,84 +728,46 @@ CellConservativeQuartic::CoarseBox (const Box&     fine,
 
 void
 CellConservativeQuartic::interp (const FArrayBox&  crse,
-                                 int               crse_comp,
+                                 int               /* crse_comp */,
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Interpolater.cpp]: We need to revert the change here because we need to use `crse_comp`.

File: Src/AmrCore/AMReX_Interpolater.cpp [hunk 2]
```diff
@@ -729,84 +728,46 @@ CellConservativeQuartic::CoarseBox (const Box&     fine,
...
                                  FArrayBox&        fine,
-                                 int               fine_comp,
+                                 int               /* fine_comp */,
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Interpolater.cpp]: We need to revert the change here because we need to use `fine_comp`.

File: Src/AmrCore/AMReX_Interpolater.cpp [hunk 3]
```diff
@@ -729,84 +728,46 @@ CellConservativeQuartic::CoarseBox (const Box&     fine,
...
-    Vector<Real> ctmp(ltmp);
-#endif
-
-#if (AMREX_SPACEDIM == 3)
-    ltmp = (cbhi[0]-cblo[0]+1)*(cbhi[1]-cblo[1]+1)*ratio[2];
-    Vector<Real> ctmp2(ltmp);
-#endif
+    // Extract pointers to fab data
+    Array4<Real const> const& crsearr = crse.const_array();
+    Array4<Real>       const& finearr = fine.array();
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Interpolater.cpp]:
```suggestion
    Array4<Real const> const& crsearr = crse.const_array(crse_comp);
    Array4<Real>       const& finearr = fine.array(fine_comp);
```
This is where `crse_comp` and `fine_comp` are needed.

File: Src/AmrCore/AMReX_Interpolater.cpp [hunk 4]
```diff
@@ -729,84 +728,46 @@ CellConservativeQuartic::CoarseBox (const Box&     fine,
...
-                       cdat,AMREX_ARLIM(clo),AMREX_ARLIM(chi),
-                       cblo, cbhi, cb2lo, cb2hi,
-                       &ncomp,
-                       AMREX_D_DECL(&ratioV[0],&ratioV[1],&ratioV[2]),
-                       AMREX_D_DECL(ftmp.dataPtr(), ctmp.dataPtr(), ctmp2.dataPtr()),
-                       bc.dataPtr(),&actual_comp,&actual_state);
+    AMREX_HOST_DEVICE_PARALLEL_FOR_4D_FLAG(runon, target_fine_region, ncomp, i, j, k, n,
+    {
+        ccquartic_interp(i, j, k, n,
+                         ratio,
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Interpolater.cpp]: Because we only support `ratio == 2`, maybe we do not have to pass `ratio`.

File: Src/AmrCore/AMReX_Interp_1D_C.H [hunk 5]
```diff
@@ -105,6 +106,35 @@ face_linear_interp_x (int i, int /*j*/, int /*k*/, int n, Array4<T> const& fine,
...
+void ccquartic_interp (int i, int /*j*/, int /*k*/, int n,
+                       IntVect const& /* ratio */,
+                       Array4<T const> const& crse,
+                       Array4<T>       const& fine ) noexcept
+
+{
+    // Note: there are asserts in CellConservativeQuartic::interp()
+    //       to check whether ratio is all equal to 2.
+
+    const Array1D<Real, -2, 2> cL = { -0.01171875,  0.0859375, 0.5, -0.0859375, 0.01171875 };
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Interp_1D_C.H]:
```suggestion
    constexpr Array1D<Real, -2, 2> cL = { -0.01171875_rt,  0.0859375_rt, 0.5_rt, -0.0859375_rt, 0.01171875_rt };
```
Make similar changes in 2d and 3d.

File: Src/AmrCore/AMReX_Interp_1D_C.H [hunk 6]
```diff
@@ -105,6 +106,35 @@ face_linear_interp_x (int i, int /*j*/, int /*k*/, int n, Array4<T> const& fine,
...
+
+    int ic = amrex::coarsen(i,2);
+    int irx = std::abs(i % 2);
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Interp_1D_C.H]: It's much cheaper to do `int irx = i - ic*2;`

File: Src/AmrCore/AMReX_Interp_1D_C.H [hunk 7]
```diff
@@ -105,6 +106,35 @@ face_linear_interp_x (int i, int /*j*/, int /*k*/, int n, Array4<T> const& fine,
     }
 }
 
+template <typename T>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Interp_1D_C.H]: I feel like we should make this non-template, because we only need the `amrex::Real` version and the function body uses `Real` in many places already.

File: Src/AmrCore/AMReX_Interp_1D_C.H [hunk 8]
```diff
@@ -105,6 +106,35 @@ face_linear_interp_x (int i, int /*j*/, int /*k*/, int n, Array4<T> const& fine,
...
+    Array1D<Real, 0, 1> ftmp;
+
+    ftmp(0) = (T)2.0 * ( cL(-2)*crse(ic-2,0,0,n)
+                       + cL(-1)*crse(ic-1,0,0,n)
+                       + cL( 0)*crse(ic  ,0,0,n)
+                       + cL( 1)*crse(ic+1,0,0,n)
+                       + cL( 2)*crse(ic+2,0,0,n) );
+    ftmp(1) = (T)2.0 * crse(ic,0,0,n) - ftmp(0);
+
+    fine(i,0,0,n) = ftmp(irx);
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Interp_1D_C.H]: There is no reason to have an array here and compute both.  We could do something like
```
Real ftmp = .....;
if (ic*2 != i) {
  ftmp = 2.0_rt * crse(ic,0,0,,n) - ftmp;
}
fine(i,0,0,n) = ftmp;
```

File: Src/AmrCore/AMReX_Interp_2D_C.H [hunk 9]
```diff
@@ -450,6 +451,46 @@ void ccprotect_2d (int ic, int jc, int /*kc*/, int nvar,
...
+
+    const Array1D<Real, -2, 2> cL = { -0.01171875,  0.0859375, 0.5, -0.0859375, 0.01171875 };
+
+    int ic = amrex::coarsen(i,2);
+    int jc = amrex::coarsen(j,2);
+    int irx = std::abs(i % 2);
+    int jry = std::abs(j % 2);
+
+    Array1D<Real,        0, 1> ftmp;
+    Array2D<Real, -2, 2, 0, 1> ctmp;
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Interp_2D_C.H]: Like in 1D, we only need `Real ftmp` and `Array1D<Real,-2,2> ctmp`.

File: Src/AmrCore/AMReX_Interp_3D_C.H [hunk 10]
```diff
@@ -617,6 +617,60 @@ void ccprotect_3d (int ic, int jc, int kc, int nvar,
...
+    int ic = amrex::coarsen(i,2);
+    int jc = amrex::coarsen(j,2);
+    int kc = amrex::coarsen(k,2);
+    int irx = std::abs(i % 2);
+    int jry = std::abs(j % 2);
+    int krz = std::abs(k % 2);
+
+    Array1D<Real,               0, 1> ftmp;
+    Array2D<Real,        -2, 2, 0, 1> ctmp;
+    Array3D<Real, -2, 2, -2, 2, 0, 1> ctmp2;
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Interp_3D_C.H]: Like in 1d and 2d, we can reduce the dimension of these by 1.  This will use less memory.

**Review by WeiqunZhang — CHANGES_REQUESTED** [review]

File: Src/AmrCore/AMReX_Interp_2D_C.H [hunk 11]
```diff
@@ -450,6 +451,48 @@ void ccprotect_2d (int ic, int jc, int /*kc*/, int nvar,
...
+                            + cL(-1)*crse(ic+ii,jc-1,0,n)
+                            + cL( 0)*crse(ic+ii,jc,  0,n)
+                            + cL( 1)*crse(ic+ii,jc+1,0,n)
+                            + cL( 2)*crse(ic+ii,jc+2,0,n) );
+        if (jry) {
+            ctmp(ii) = 2.0_rt * crse(ic+ii,jc,0,n) - ctmp(ii);
+        }
+    } // ii
+
+    ftmp = 2.0_rt * ( cL(-2)*ctmp(-2)
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Interp_2D_C.H]: Minor suggestion.  `Real ftmp = ...` instead of
```
Real ftmp;
// many lines of code
ftmp = ...
```

Comment (wyphan) [reply Src/AmrCore/AMReX_Interp_2D_C.H]: I guess this has something to do with my Fortran background 😂

**Review by WeiqunZhang — APPROVED** [review]
